Automatic CSR generation - first steps

LiptonB opened the pull request.

LiptonB commented:
The new/modified LDAP objects created by this change can be viewed in [newentries.txt](https://github.com/LiptonB/freeipa/files/335488/newentries.txt)

MartinBasti commented on ipaserver/plugins/certmapping.py:
```diff
@@ -239,11 +248,186 @@ class cert_get_requestdata(Command):
+
+        return config
+
+class CertutilHelper(object):
```
If you want to have common interface there, you should use abstract class and abstract method (format) and inherit it in CerutilHelper and OpenSSLHelper

LiptonB replied:
Updated.

MartinBasti commented on ipaserver/plugins/certmapping.py:
```diff
@@ -239,11 +248,186 @@ class cert_get_requestdata(Command):
+    SECTION_RE = re.compile(r'%section\((.*?)\)', re.DOTALL)
+    EXTENSION_RE = re.compile(r'^ext:(.*)$', re.DOTALL+re.MULTILINE)
+
+    def parse_sections(self, value, sections):
```
This should be protected or even private method, use name _process_section or __process_section

LiptonB replied:
Good point. Changed to _parse_sections.

MartinBasti commented on ipaserver/plugins/certmapping.py:
```diff
@@ -239,11 +248,186 @@ class cert_get_requestdata(Command):
+        if 'subject' in principal:
+            subject = principal['subject']
+        else:
+            _, principal_name, _ = split_any_principal(principal['krbprincipalname'][0])
```
Please don't use '_' we use it as alias for ugettext in the whole code, just use variable name with prefix '_' (we will use this notation in future)

LiptonB replied:
Done. Well, actually this code just disappeared because of the new way principals are handled.

MartinBasti commented on ipaserver/plugins/certmapping.py:
```diff
@@ -239,11 +248,186 @@ class cert_get_requestdata(Command):
+
+    def dataDNSOpenssl(self, extra_inputs):
+        principal = extra_inputs['principal']
+        _, principal_name, _ = split_any_principal(principal['krbprincipalname'][0])
```
_: same here

MartinBasti commented on ipaserver/plugins/certmapping.py:
```diff
@@ -239,11 +248,186 @@ class cert_get_requestdata(Command):
+
+    def dataDNSCertutil(self, extra_inputs):
+        principal = extra_inputs['principal']
+        _, principal_name, _ = split_any_principal(principal['krbprincipalname'][0])
```
_: same here

MartinBasti commented on ipaserver/install/cainstance.py:
```diff
@@ -1750,27 +1754,68 @@ def import_included_profiles():
     api.Backend.ra_certprofile.override_port = 8443
 
-    for (profile_id, desc, store_issued) in dogtag.INCLUDED_PROFILES:
+    for (rule_id, template, helper) in dogtag.INCLUDED_TRANSFORMATION_RULES:
```
can you use API command here you introduced in previous commit?

LiptonB replied:
I'm not sure. I can't tell whether it's possible to make API calls during the install/upgrade process. I'm currently getting the following error, but I may be doing something wrong:

```
  File "/usr/lib/python2.7/site-packages/ipaserver/install/service.py", line 448, in start_creation
    run_step(full_msg, method)
  File "/usr/lib/python2.7/site-packages/ipaserver/install/service.py", line 438, in run_step
    method()
  File "/usr/lib/python2.7/site-packages/ipaserver/install/cainstance.py", line 1837, in import_included_profiles
    api.Command.certtransformationrule_show(rule_id)
  File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 447, in __call__
    return self.__do_call(*args, **options)
  File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 475, in __do_call
    ret = self.run(*args, **options)
  File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 797, in run
    return self.execute(*args, **options)
  File "/usr/lib/python2.7/site-packages/ipaserver/plugins/baseldap.py", line 1318, in execute
    dn, attrs_list
  File "/usr/lib/python2.7/site-packages/ipaserver/plugins/baseldap.py", line 1098, in wrapped
    return func(*call_args, **call_kwargs)
  File "/usr/lib/python2.7/site-packages/ipapython/ipaldap.py", line 1477, in get_entry
    size_limit=size_limit
  File "/usr/lib/python2.7/site-packages/ipapython/ipaldap.py", line 1301, in get_entries
    base_dn=base_dn, scope=scope, filter=filter, attrs_list=attrs_list)
  File "/usr/lib/python2.7/site-packages/ipapython/ipaldap.py", line 1346, in find_entries
    time_limit = self.time_limit
  File "/usr/lib/python2.7/site-packages/ipaserver/plugins/ldap2.py", line 84, in time_limit
    return float(self.get_ipa_config().single_value.get(
  File "/usr/lib/python2.7/site-packages/ipaserver/plugins/ldap2.py", line 240, in get_ipa_config
    time_limit=2, size_limit=10
  File "/usr/lib/python2.7/site-packages/ipapython/ipaldap.py", line 1378, in find_entries
    id = self.conn.search_ext(
  File "/usr/lib/python2.7/site-packages/ipalib/backend.py", line 105, in __get_conn
    threading.currentThread().getName()
AttributeError: ldap2 is not connected (ldap2_140357588210704 in MainThread)
```

Sort of a separate issue, but I'm working on code that will allow profiles with field mapping rules to be imported from files (and at some point will probably do the same for mapping rulesets/transformation rules). This might allow some of the details of the rules to be taken out of the initialization code and moved to profile files instead, which would be nice. That might also depend on the API working, though.

LiptonB committed:
Add schema to support automatic CSR generation

This adds the schema discussed in
http://www.freeipa.org/page/V4/Automatic_Certificate_Request_Generation/Schema#Option_A, along with containers for the three new objectClasses.

LiptonB committed:
Add plugin for CSR generation

This plugin will implement the cert-get-requestdata call that returns a
config that can be used to generate a CSR. This commit implements only
the very basic API of the plugin. Actual functionality will come in
later patches.

LiptonB committed:
Add generation rules to the default cert profile

This updates the automation that happens on upgrade to create all the
various types of mapping rules, and adds some default rules to the
caIPAserviceCert profile.

LiptonB committed:
Add some actual transformation rules

For now, these rules are implemented in python for simplicity, and
referenced by naming the method prefixed with 'py:' in the
transformation template attribute, but eventually we aim to define a
text-based syntax that can be used in the attribute directly and thus
modified at runtime.

LiptonB commented:
Thanks for the review, @mbasti-rh! I've rebased the code and addressed/responded to most of your comments. Please let me know if you see anything else that looks amiss.

LiptonB committed:
Make field mappings optional

LiptonB committed:
Allow multiple helpers per rule

LiptonB committed:
Add a new cert profile for users

LiptonB committed:
Add ability to import/export mappings with profile

The certprofile_show command now takes a --mappings-out flag that writes
mappings to a JSON-formatted file, and the same file format can be read
in with the --mappings-file parameter of certprofile_import and
certprofile_mod.

LiptonB committed:
Make helper parameter multivalued to match schema

LiptonB committed:
Add tests for mapping rules import/export

Also fix a couple of small bugs exposed by the tests.

LiptonB committed:
Clean up exception handling in certprofile_import

* If None is passed for 'mappings_file', ignore that parameter rather
  than throwing an internal error.
* Since we have an exc_callback now, overriding execute is unnecessary,
  so move that exception handling into the callback as well.

LiptonB committed:
Small test improvements

* The order of the mappings returned by certprofile_show is not
  significant
* Assert that LDAP entries for malformed profiles don't stick around
  after the import call fails

LiptonB commented:
Responding to a comment from @mbasti-rh, I now have an experiment working that uses the API to create the included mapping rules instead of creating them directly with LDAP. I think it's a good approach, but it turned out to touch a bunch of things so I'm leaving it separate for now. That code is here: https://github.com/LiptonB/freeipa/pull/3
